# 1.2 手続きとその生成するプロセス

・**プロセス** とは、一つの仕事を表す単位、または処理そのもののことである。（時田解釈）  
・プロセスの **局所的展開(local evolution;書籍では局所的進化)** は、手続きによって記述される。  
・手続きの全体的なふるまい、**グローバル(global;書籍では大局的)**なふるまいについて何か記述できるようにしたい。  

本節では、単純な手続きによって作られたプロセスの、いくつかの形を見ていく。  
具体的には、  
・線形再帰、反復、木の再帰(木構造再帰)・・・空間(スペース)の消費  
・増加オーダー(増加の程度)・・・時間の消費  
である。

## 1.2.1 線形再帰と反復

階乗関数について考える。定義は以下の通り。  
$ n!=n \cdot (n-1) \cdot (n-2) \cdots 3 \cdot 2 \cdot 1 $  

階乗の計算方法として、$ (n-1)! $を計算し、その計算結果に$ n $をかけることで$ n! $を計算できる。  
$ n!=n \cdot [(n-1) \cdot (n-2) \cdots 3 \cdot 2 \cdot 1]=n \cdot (n-1)! $  

これに$ 1!=1 $という規定を加えると、そのまま手続きとして実装できる。

In [1]:
(define (factorial n)
  (if (= n 1) 1
      (* n (factorial (- n 1))
         )
      )
  )

In [2]:
(factorial 6)

720

この手続きが$ 6! $の計算を実行する様子を見るには、1.1.5節の置換モデルを使うことができる。

### 図1.3 $ 6! $を求めるための線形再帰プロセス

    (factorial 6)
    (* 6 (factorial 5))
    (* 6 (* 5 (factorial 4)))
    (* 6 (* 5 (* 4 (factorial 3))))
    (* 6 (* 5 (* 4 (* 3 (factorial 2)))))
    (* 6 (* 5 (* 4 (* 3 (* 2 (factorial 1))))))
    (* 6 (* 5 (* 4 (* 3 (* 2 1)))))
    (* 6 (* 5 (* 4 (* 3 2))))
    (* 6 (* 5 (* 4 6)))
    (* 6 (* 5 24))
    (* 6 120)
    720


In [3]:
;階乗の計算の様子を出力させてみた。
(define (factorial2 n)
  (if (= n 1)
      (begin
       (display n)
       (display '!=)
       (display n)
       (newline)
       1)
       (begin
        (display n)
        (display '!=)
        (display n)
        (display '* )
        (display (- n 1))
        (display '!)
        (newline)
        (define ret (* n (factorial2 (- n 1))))
        (display n)
        (display '!=)
        (display ret)
        (newline)
        ret
        )
      )
  )

In [4]:
(factorial2 6)

6!=6*5!
5!=5*4!
4!=4*3!
3!=3*2!
2!=2*1!
1!=1
2!=2
3!=6
4!=24
5!=120
6!=720


720

In [5]:
;階乗の計算の様子を出力させてみた。
;上記だと(define ret ・・・)でエラーになる場合があった。
(define (factorial2 n)
  (if (= n 1)
      (begin
       (display n)
       (display '!=)
       (display n)
       (newline)
       1)
        (begin
          (display n)
          (display '!=)
          (display n)
          (display '* )
          (display (- n 1))
          (display '!)
          (newline)
          (let ((ret (* n (factorial2 (- n 1)))))
            (begin
              (display n)
              (display '!=)
              (display ret)
              (newline)
              ret
            )
          )
        )
    )
  )

In [6]:
(factorial2 6)

6!=6*5!
5!=5*4!
4!=4*3!
3!=3*2!
2!=2*1!
1!=1
2!=2
3!=6
4!=24
5!=120
6!=720


720

階乗の別の計算方法を考えてみる。  
$ n! $の計算規則として、最初に$ 1 \times 2 $を行い、それからその答えに$ 3 $をかけ、それから$ 4 $をかけ、$ n $まで続けるというものである。  
形式的に書くと、計算途中の積と$ 1 $から$ n $まで数え上げていくカウンタを保持することで、次の規則に従ってカウンタと積がステップごとに同時に変化するように記述できる。 
$ product \leftarrow counter \times product $  
$ counter \leftarrow counter + 1 $  

$ n! $とはカウンタが$ n $を超えた時点での積の値であると規定すると、以下のように実装できる。

In [7]:
(define (factorial3 n)
  (fact-iter 1 1 n))
(define (fact-iter product counter max-count)
  (if (> counter max-count)
      product
      (fact-iter (* counter product) (+ counter 1) max-count)
    )
  )

In [8]:
(factorial3 6)

720

先ほどと同じ悪、置換モデルを使い、視覚化すると以下のようになる。

### 図1.4 $ 6! $を計算する線形反復プロセス

    (factorial 6)
    (fact-iter   1 1 6)
    (fact-iter   1 2 6)
    (fact-iter   2 3 6)
    (fact-iter   6 4 6)
    (fact-iter  24 5 6)
    (fact-iter 120 6 6)
    (fact-iter 720 7 6)


In [9]:
;階乗の計算の様子を出力させてみた。
(define (factorial4 n)
  (fact-iter2 1 1 n))
(define (fact-iter2 product counter max-count)
  (if (> counter max-count)
      (begin
       product
       )
      (begin
       (display "fact-iter2 ")
       (display product)
       (display " ")
       (display counter)
       (display " ")
       (display max-count)
       (newline)
       (fact-iter2 (* counter product) (+ counter 1) max-count)
       )
    )
  )

In [10]:
(factorial4 6)

fact-iter2 1 1 6
fact-iter2 1 2 6
fact-iter2 2 3 6
fact-iter2 6 4 6
fact-iter2 24 5 6
fact-iter2 120 6 6


720